### DISCUSSION TITLE
UBIQUITOUS LANGUAGE

### DISCUSSION THREAD

**Discussion by eastshine-high** [CodeSoom/DDD-Kurly-Clone-Promotion#7] [General]
### 할인(discount)
* 등록된 상품에 정률 할인을 등록한다(Register a flat-rate discount on a registered product)
    * 정률 할인(flat-rate discount)
    * 할인을 등록할 수 있다(You can register for a discount)
* 할인중인 상품(products on discount)

---
## 피드백을 반영하여 아래에 글을 추가했습니다.

**Comment by etff** [comment]
할인의 도메인입장에서 유비쿼터스 언어를 정의해보면 어떨까요?
ex) 
1. 할인을 상품에 적용할 수 있다.
2. 할인은 정액 할인 정책을 가지고 있다.

**Reply by etff** [reply]
유비쿼터스 언어의 크게 소리내어 말하기는 우리가 소리내어 말했을 때 말이되는지 검증하는 과정입니다.
``` 등록된 상품에 정액 할인을 등록한다 ``` 여기서 정액 할인을 등록한다는 조금 어색하다고 생각하는데 어떻게 생각하시나요?

**Reply by daadaadaah** [reply]
1번 관련해서, "할인은 상품에 적용될 수 있다." 는 어떠신가요?

**Reply by etff** [reply]
```할인중인 상품(products on discount) ``` 명사로 표현해보면 어떨까요?

**Reply by eastshine-high** [reply]
좋은 의견입니다! 할인의 도메인 입장에서 언어를 정의하는 것이 좋아보이네요.

**Reply by daadaadaah** [reply]
> `할인중인 상품(products on discount) ` 명사로 표현해보면 어떨까요?

그냥 "할인 상품"이라고 해도 괜찮을 것 같다는 생각이 듭니다ㅎㅎ

**Reply by etff** [reply]
저희가 해야할 일중 이번에 할인 적용된 상품을 조회하는 것도 포함되어야할까요?

**Reply by eastshine-high** [reply]
저는 `할인은 상품에 적용할 수 있다` 능동 표현으로 생각했는데 수동 표현일까요?

**Reply by etff** [reply]
이미 저희 스터디때도 소개했는데요. 

유비쿼터스 랭귀지를 어떻게 작성하는지어렵다면, 스포카에서 적용중인 예제를 참고해보면 어떨까요?
https://spoqa.github.io/2021/09/13/domain-driven-development-transition-story.html

**Reply by etff** [reply]
```등록된 상품```은 정의를 하셨을까요? ```등록된``` 상품이라는 표현도 고민해보시기 바랍니다.

**Reply by etff** [reply]
위 링크를 보시고 속성의 정의도 해보시길 바랍니다

**Reply by eastshine-high** [reply]
네! 피드백 감사합니다.

**Reply by daadaadaah** [reply]
> 저희가 해야할 일중 이번에 할인 적용된 상품을 조회하는 것도 포함되어야할까요?

FE팀에서는 "메인 페이지에서 할인된 단일 상품을 보인다"라는 기능을 이번주 TODO에 있긴 해요!

**Reply by daadaadaah** [reply]
> `등록된 상품`은 정의를 하셨을까요? `등록된` 상품이라는 표현도 고민해보시기 바랍니다.

문득 등록되지 않는 상품도 있나? 라는 생각이 들었어요!ㅎㅎ
"그냥 상품이라도 표현해도 괜찮지 않을까" 라는 생각이 듭니다.
혹시, 등록된 상품과 등록되지 않는 상품을 나눈 기준이 "일반 사용자 홈페이지에서 볼 수 있는지 여부"라면, 필요하겠다는 생각도 드네용ㅎㅎ

**Reply by daadaadaah** [reply]
2번 관련해서, "정책" 보다 "방식 또는 방법" 어떠신가요?
정책이란 용어는 예를 들어, 중복 할인 불가 등 이런 rule 적인 요소에 사용하기 위해 아껴두는것도 좋은 방법 같아요!ㅎㅎ혹시, 다른 의견 있으시면 답글 남겨주세요~ 프로모션 도메인은 처음이라 잘 몰라요😂

**Reply by eastshine-high** [reply]
정책이란 표현을 전부 사용하는 건 어떤가요? 정액 할인에도 정책이 있을 것 같기는 해서요. 예를 들어 "상품의 금액 이상 정액 할인을 할 수 없다."

**Reply by etff** [reply]
제가 생각하는 ```정책```은 핸드폰 요금제처럼 청소년할인, 야간할인, 커플할인 같은 목표를 달성하는 방법을 말합니다.
일정 금액 사용, 시간제한, 중복 적용 불가 같은 것은 정책의  ```조건(condition)```이라고 생각하는데 다른분들은 어떻게 생각하시는지 궁금하네요.

**Reply by eastshine-high** [reply]
도메인 주도 책 P.18에 있는 예제에서는 조건절을 정책으로 리팩토링하는 것 같더라구요. 그래서 저는 @daadaaadaah 처럼 <del> rule 적인 요소로 이해하고 있었습니다.</del>

**Reply by eastshine-high** [reply]
제가 이해한 것을 서술하면 @etff님 말씀처럼 서술될 것 같네요!

**Reply by daadaadaah** [reply]
> 제가 생각하는 `정책`은 핸드폰 요금제처럼 청소년할인, 야간할인, 커플할인 같은 목표를 달성하는 방법을 말합니다. 일정 금액 사용, 시간제한, 중복 적용 불가 같은 것은 정책의 `조건(condition)`이라고 생각하는데 다른분들은 어떻게 생각하시는지 궁금하네요.

오!! '조건'이란 용어로 뺄 수 있겠네용!!
그럼, 저는 '정책' 용어 사용하는거 동의합니다👍

**Comment by eastshine-high** [comment]
### 할인(Discount)
* `할인`을 `상품`에 등록할 수 있다.
    * `할인` 대상 `상품`을 정할 수 있다.
    * `할인`은 `정률 할인`을 가진다.
    * `할인`은 `정액 할인`을 가진다.
    * `할인`은 `시작일`과 `만료일`의 기간을 가진다.
* `할인`을 `종료`할 수 있다.
    * `할인`은 관리자만 종료할 수 있다.
* `할인`을 `조회`할 수 있다. 
* `할인`을 `수정`할 수 있다. 

### Attribute
Discount
| 속성 | 설명             | 관계 |
| ------- | --------- | -------- |
| id   | 할인의 식별자 |  |
| flatRate   | 정률(할인 비율) |  |
| fixedAmound | 정액(할인 금액) |  |
| startDate | 할인 시작일 |  |
| expiryDate | 할인 종료일 |  |
| productId | 상품의 식별자 |  |

**Reply by etff** [reply]
저희는 정액 할인을 합의했던 것같은데, ```정률```이 되었네요. 혹시 이렇게 변경된 이유를 알 수 있을까요?
정액할인은 amount discount로 일정 값을 말하는 반면, rate가 들어가면 비율 할인로 이해할 가능성이 높은데 의도된 것일까요?

**Reply by etff** [reply]
정액할인은 @daadaadaah 님이 올려준 쿠팡의 정의에 따르면
```정액할인``` 총 구매 금액에서 설정한 설정한 할인 금액을 차감하는 방식입니다.
로 나오네요.

**Reply by etff** [reply]
> 할인을 상품에 등록할 수 있다. (Discounts can be registered on products)

추가적으로 영어표현을 쓰실 필요는 없을 것 같습니다.

**Reply by eastshine-high** [reply]
네 디스코드에 @daadaaadaah 님 요청을 반영해서 수정하였습니다.
이번 주차는 정률(rate) 할인을 구현하고 할인 방식(정률, 정액)에 대해서는 세부적으로 추가 논의하면 좋을 것 같습니다.
영문 표현은 제거하겠습니다.

**Reply by daadaadaah** [reply]
> 네 디스코드에 @daadaaadaah 님 요청을 반영해서 수정하였습니다. 이번 주차는 정률(rate) 할인을 구현하고 할인 방식(정률, 정액)에 대해서는 세부적으로 추가 논의하면 좋을 것 같습니다. 영문 표현은 제거하겠습니다.

위의 디스커션에도 정률로 변경해야 할 것 같아요!ㅎㅎ

**Reply by etff** [reply]
별거아니지만 정률이라는 용어가 도메인전문가(현업)가 쓰는 용어인지 체크해보는것도 좋을 것같네요

**Reply by etff** [reply]
정률은 flat rate가 나을까요 fixed rate가 나을까요? 🤔

**Reply by eastshine-high** [reply]
저는 flat rate에 한 표 하겠습니다. https://budgeting.thenest.com/backtoback-swaps-32402.html 
참조한 글의 설명에서는 단어의 사용 용도가 조금 다르다고 설명하네요. fixed rate은 이율과 관련된 곳에서 사용하는 것 같고, flat rate은 비용 청구와 관련할 때 사용하는 것 같아요.

**Reply by eastshine-high** [reply]
@daadaadaah  네! 수정했습니다.

**Reply by etff** [reply]
> 저는 flat rate에 한 표 하겠습니다. https://budgeting.thenest.com/backtoback-swaps-32402.html 

좋네요! 용어가 다르군요!

**Comment by etff** [comment]
### 쿠폰(Coupon)
* `쿠폰`을 `관리자`가 `등록`할 수 있다.
    *  `쿠폰`은 `쿠폰번호`를 가진다.
    * `쿠폰`은 `정률 할인` 정책을 가진다.
    * `쿠폰`은 `정액 할인` 정책을 가진다.
    * `쿠폰`은 `시작일`과 `만료일`의 `기간`을 가진다.
    * `쿠폰`의 `최소 적용 금액` 조건을 설정할 수 있다.
* `쿠폰`을 만들 개수를 지정할 수 있다.
* `쿠폰`을 `사용자`에게 `발급` 할 수 있다.
  * `쿠폰번호`를 `사용자`가 입력하면 `발급`된다.
  * `쿠폰번호`가 존재하지 않는 번호를 `사용자`가 입력하면 오류가 발생한다.
  * `쿠폰`이 `발급`되면 `사용여부`를 가진다.
* `쿠폰` 목록을 가져올 수 있다.
    * `쿠폰`을 `사용자`는 자신의 것만 볼 수 있다.
    * `쿠폰`을 `사용자`는 `만료`된 것은 볼 수 없다.
    * `쿠폰`을 `관리자`는 모두 볼 수 있다.
* `쿠폰`을 `주문`에 적용할 수 있다.
* `쿠폰`을 `만료`할 수 있다. 
  * `쿠폰`이 `만료` 되면 사용할 수 없다.

### Attribute
Coupon
| 속성 | 설명             | 관계 |
| ------- | --------- | -------- |
| id   | 쿠폰의 식별자 |  |
|사용자 | 개인 정보를 보유하고 있으며 물건 구매를 위한 인증된 사용자| |
|관리자 | 쿠폰을 관리하고 시스템을 관리하는 사용자| |
| 쿠폰번호   | 쿠폰을 사용자가 등록할때 쓰는 번호 |  |
| 정률   | 쿠폰의 적용 비율 |  |
| 정액   | 적용 금액 |  |
| 개수   | 사용가능한 쿠폰의 총 개수 |  |
| orderId | 주문의 식별자 |  |
| 기간 | 쿠폰이 적용 가능한 기간 |  |
| 만료 | 쿠폰이 종료되어 사용할 수 없는 상태 |  |
| 사용여부 | 쿠폰 사용여부 |  |

---
## 쿠폰 도메인을 정의했습니다

**Reply by hannut91** [reply]
쿠폰의 상태를 가지고 있는 것은 어떨까요?

생성됨
발급됨
사용됨
만료됨

이러면 발급은 되었지만, 사용하지 않고 만료된 것은 구분할 수가 없어서 별로일까요 흠

**Reply by hannut91** [reply]
* `쿠폰`을 `관리자`가 `등록`할 수 있다.
    *  `쿠폰`은 `쿠폰번호`를 가진다.
    * `쿠폰`은 `정률 할인` 정책을 가진다.
    * `쿠폰`은 `정액 할인` 정책을 가진다.
    * `쿠폰`은 `시작일`과 `만료일`의 `기간`을 가진다.
* `쿠폰`을 만들 개수를 지정할 수 있다.

쿠폰을 만들 개수를 지정할 수 있는 것은 관리자가 할 수 있는 것이니까, 인덴트를 이렇게 해야겠네요

**Reply by hannut91** [reply]
> 쿠폰을 사용자는 만료되거나 사용된 것은 볼 수 없다.

사용자는 사용한 것을 볼 수 있어야 할 것 같아요. 그런데 이걸 언제까지 보여줄지도 정해야 할 것 같네요

![image](https://user-images.githubusercontent.com/14071105/137575726-01532b0c-b5fd-47b1-b61b-2eefb006fcec.png)

아니 근데 보니까 사용 가능이 2장으로 나오네... 버그발견...

**Reply by etff** [reply]
뜻하지않은 qa 발생 😭

**Reply by etff** [reply]
쿠폰번호(coupon number) 용어가 적절할지 고민이네요. 🤔

쿠폰코드(coupon code)가 더 나은 용어일까요?

**Reply by etff** [reply]
```관리자```가 등록하는 것은 ```쿠폰 정책```이고, ```사용자```에게 실질적으로 발급되는 것은 ```쿠폰```으로 변경하는 것은 어떨까요?

**Reply by etff** [reply]
쿠폰을 만드는 것과 쿠폰을 쓸 때 용어를 분리해야될 것같네요. 😄

**Reply by eastshine-high** [reply]
> 쿠폰을 만드는 것과 쿠폰을 쓸 때 용어를 분리해야될 것같네요. 😄

그러네요. `발행(published or Released) 쿠폰` 과 `발급(issued) 쿠폰` 처럼 구분할 필요가 있어보여요.
https://forum.wordreference.com/threads/publish-issue-release.2632953/#post-13279471

**Reply by eastshine-high** [reply]
> 쿠폰번호(coupon number) 용어가 적절할지 고민이네요. 🤔
> 
> 쿠폰코드(coupon code)가 더 나은 용어일까요?

저는 `쿠폰 코드`가 더 의미가 분명한 것 같기도 하네요? 실제로 coupon code라는 말을 사용하는 것 같기도 하네요.
https://docs.magento.com/user-guide/marketing/price-rules-cart-create.html

**Reply by etff** [reply]
> > 쿠폰번호(coupon number) 용어가 적절할지 고민이네요. 🤔
> > 쿠폰코드(coupon code)가 더 나은 용어일까요?
> 
> 저는 `쿠폰 코드`가 더 의미가 분명한 것 같기도 하네요? 실제로 coupon code라는 말을 사용하는 것 같기도 하네요. https://docs.magento.com/user-guide/marketing/price-rules-cart-create.html

```쿠폰 코드```가 더 명확할 것같네요!

**Reply by etff** [reply]
> > > 쿠폰번호(coupon number) 용어가 적절할지 고민이네요. 🤔
> > > 쿠폰코드(coupon code)가 더 나은 용어일까요?
> > 
> > 
> > 저는 `쿠폰 코드`가 더 의미가 분명한 것 같기도 하네요? 실제로 coupon code라는 말을 사용하는 것 같기도 하네요. https://docs.magento.com/user-guide/marketing/price-rules-cart-create.html
> 
> `쿠폰 코드`가 더 명확할 것같네요!

회의결과 사용자 번호 입력시 쿠폰이 등록되는 용어는 `키워드`로 정해졌습니다. 위키에 반영하였습니다.
